### ISSUE TITLE
[no-shadow] False positive in function block as variable is initialized

### ISSUE THREAD

**Issue by binoche9** [eslint/eslint#13943]
**Tell us about your environment**

* **ESLint Version:**: 7.16.0
* **Node Version:**: [ESLint Demo](https://eslint.org/demo)
* **npm Version:**: [ESLint Demo](https://eslint.org/demo)

**What parser (default, `@babel/eslint-parser`, `@typescript-eslint/parser`, etc.) are you using?**

default parser, which I assume is what the ESLint Demo uses.

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  rules: {
    "no-shadow": "error",
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const foo = func(() => {
  const foo = 3;
  return foo + 2;
});
```

Repro in [ESLint Demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLXNoYWRvdzogWydlcnJvciddICovXG5cbmNvbnN0IGZvbyA9IGZ1bmMoKCkgPT4ge1xuICBjb25zdCBmb28gPSAzO1xuICByZXR1cm4gZm9vICsgMjtcbn0pOyIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6MTEsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

**What did you expect to happen?**

No error should be thrown because inside the function block the outer foo hasn't been initialized yet, so it shouldn't be in scope. I hoped that the `{ "hoist": "never" }` option would help but sadly it didn't.

**What actually happened? Please include the actual, raw output from ESLint.**

The lint rule thinks that the inner foo is shadowing the outer foo.

```
2:9 - 'foo' is already declared in the upper scope.
```

**Are you willing to submit a pull request to fix this bug?**

Yes, with some pointers on where to start (especially if it involves the scope manager)

**Comment by mdjermanovic** [comment]
Hi @binoche9, thanks for the issue!

I think this is a duplicate of https://github.com/eslint/eslint/issues/12687, which is already accepted.

> No error should be thrown because inside the function block the outer foo hasn't been initialized yet, so it shouldn't be in scope. I hoped that the `{ "hoist": "never" }` option would help but sadly it didn't.

The outer `foo` is already in scope and can be even initialized at the time when the function will be called, so this is shadowing but it's arguable should this rule report it if `hoist` is set to `"functions"` (default) or `"never"`.

**Comment by binoche9** [comment]
Ah you're right that it seems like a duplicate. There even is an open [PR](https://github.com/eslint/eslint/pull/13688) (that you reviewed) for it that seems to have been neglected by the author. Would it be okay to re-open a new one?

**Comment by mdjermanovic** [comment]
I believe the PR isn't abandoned but is waiting for the open question about the additional option.

> I also liked to reply to some things you said in a comment in that PR - i guess it would be best to write it there? (just worried that the PR seems a bit abandoned)

Sure! PR seems like a good place to continue the discussion.

**Comment by binoche9** [comment]
> I believe the PR isn't abandoned but is waiting for the open question about the additional option.

Just curious, when is a PR deemed officially abandoned? I ask this because it seems like the author hasn't responded since the PR was opened 3 months ago.

**Comment by mdjermanovic** [comment]
When the author doesn't respond to change requests in 30 days, though we usually first ask if they would like to continue before closing the PR. Regarding this PR, there wasn't a change request, it looked to me that it isn't yet clear what exactly should be done.

**Comment by binoche9** [comment]
I changed my mind on replying in the PR. I thought the outer variable would be un-referenceable from the inner function even if it's executed later, but I realized I was testing it wrong. So I now agree it should be a new option if anything, but I don't feel strongly anymore because I now realize it _is_ shadowing.
